### ISSUE TITLE
Curl cause segment fault when inittate HTTPS on SHA_init of libcrypto

### ISSUE THREAD

**Issue by V-VnV** [curl/curl#10166]
### I did this

In our program we need to send HTTPS message (thread send), we separate send message to 2 steps: handshake and handle data.
As my understanding, handshake is first step and it require encryption for certificate and CA. There are option we used for it:
curl_easy_setopt(curl_post, CURLOPT_SSL_VERIFYPEER, 0);
curl_easy_setopt(curl_post, CURLOPT_SSL_VERIFYHOST, 0);
curl_easy_setopt(curl_post, CURLOPT_SSLCERTTYPE, "PEM");
curl_easy_setopt(curl_post, CURLOPT_SSLCERT, pCertFile);
curl_easy_setopt(curl_post, CURLOPT_CAPATH, pCACertPath);
curl_easy_setopt(curl_post, CURLOPT_CAINFO, pCACertFile);
curl_easy_setopt(curl_post, CURLOPT_SSLKEYTYPE, "PEM");

For handle data we used these option:
CURLOPT_URL
CURLOPT_INTERFACE
CURLOPT_PORT
CURLOPT_POST = 1
CURLOPT_HEADER = 1
CURLOPT_HTTPHEADER
CURLOPT_POSTFIELDS
CURLOPT_POSTFIELDSIZE
CURLOPT_TIMEOUT = 5
CURLOPT_CONNECTTIMEOUT = 5
CURLOPT_NOSIGNAL = 1

After set all of these options, curl_easy_perform() crashed.
In backtrace we found it was crashed about SHA_init:
#3  [<0x7fa3c0c6c0>] (<Unknown> + 0x0) [[vdso] + 0x6c0]
------ Cut Out Signal Frame, Actual backtrace is in below ------
#4  [<0x7fa368bd04>] (SHA_Init + 0x364) [/usr/lib64/libcrypto.so.1.0.0 + 0x79d04]

In our program, we don't have any usage about SHA_init, and that happen at same thread that use curl_easy_perform().
Additionally, we have an other thread (2) to get status from server with same handshake options and it was executed before send HTTPS message when crash happen.
 Almost options for thread (2) are same as thread send HTTPS, some different are:
CURLOPT_HTTPGET instead CURLOPT_POST
CURLOPT_TIMEOUT = 70
CURLOPT_WRITEFUNCTION
CURLOPT_WRITEDATA

### I expected the following

Program doesn't crash

### curl/libcurl version

curl_version_info() = 7.75.0

### operating system

Linux 4.4.285-yocto-standard

**Comment by bagder** [comment]
I would urge you to provide a means for us to reproduce the problem. Otherwise we can only offer you some hints on how you can debug this.

Also, you did not include OpenSSL version and you did not show enough back trace of curl to understand where or how this crash came to be.

I do not recognize this as an old known problem.

Your repeated mention of threads makes me want to make sure you [play by the rules](https://curl.se/libcurl/c/threadsafe.html).

**Comment by bagder** [comment]
I have failed to reproduce this problem myself, so can I ask you to provide us with more details on how to go ahead and repeat this. Preferably with a command line or a stand-alone program we can run from our ends against a public URL to trigger the problem?

Without more info, we have nothing to go on and this issue will be closed.

**Comment by V-VnV** [comment]
I'm sorry, I just return from a trip so I can't reply you soon.

This is our customer's issue and the program was handed over to me recently so it really hard to debug this issue due testing environment and the origin developer was left.

Anyway, I looked again about how program use CURL, and I saw something:
- Our program have two threads use CURL, but it doesn't global initial. And the first initial thread cleans its CURL handler quite much.
- Callback function for CURLOPT_WRITEFUNCTION, we let it doesn't return result in some case.

I don't know if above thing could affect to this issue or not. But I submitted changed for modifying usage of CURL, hopeful it could fix this error.

For SSL version we use version:
OpenSSLDie
OpenSSL 1.0.2h  3 May 2016

Unfortunately, our backtrace just support mostly for our program. External library live CURL, OpenSSL, .. it can show some function was use only.
I think it is OK to close this ticket because lack of infomation.
If I have more detail for this issue I'll ask for help late.

Thank you.
